### PR TITLE
ACL V2 draft implementation

### PR DESCRIPTION
**Pull request by madolson**
A Draft PR to play around with while I build up the ACL v2 code.

### CODE REVIEW

File: src/acl.c [hunk 1]
```diff
@@ -253,21 +369,22 @@ user *ACLCreateUser(const char *name, size_t namelen) {
...
     listSetMatchMethod(u->passwords,ACLListMatchSds);
     listSetFreeMethod(u->passwords,ACLListFreeSds);
     listSetDupMethod(u->passwords,ACLListDupSds);
-    listSetMatchMethod(u->patterns,ACLListMatchSds);
-    listSetFreeMethod(u->patterns,ACLListFreeSds);
-    listSetDupMethod(u->patterns,ACLListDupSds);
-    listSetMatchMethod(u->channels,ACLListMatchSds);
-    listSetFreeMethod(u->channels,ACLListFreeSds);
-    listSetDupMethod(u->channels,ACLListDupSds);
-    memset(u->allowed_commands,0,sizeof(u->allowed_commands));
```
Comment (hpatro) [inline src/acl.c:270]: Do we want to completely remove the existing mechanism ? I believe we should still support the old way of user creation and help the user move to the new mechanism.

Also can we modify a user's existing rules to a single selector ?

File: src/acl.c [hunk 2]
```diff
@@ -244,6 +266,100 @@ void *ACLListDupSds(void *item) {
...
+    UNUSED(a); UNUSED(b);
+    return 0;
+}
+
+/* TODO: Actually write the code to deuplicate */
+void *ACLListDuplicateSelector(void *src) {
+    aclSelector *new = ACLCreateSelector(0);
+    ACLCopySelector(new, src);
+    return new;
+}
```
Comment (hpatro) [inline src/acl.c:353]: I believe we have to leave it to the user/admin management to keep the selectors list clean

e.g. selector(s) having subset keyspace but same commands?

File: src/acl.c [hunk 3]
```diff
@@ -640,34 +746,75 @@ sds ACLDescribeUserCommandRules(user *u) {
...
+    } else {
+        listRewind(selector->patterns,&li);
+        while((ln = listNext(&li))) {
+            keyPattern *thispat = (keyPattern *)listNodeValue(ln);
+            if (thispat->flags == (CMD_KEY_READ | CMD_KEY_WRITE)) {
+                res = sdscatlen(res,"~",1);
+            } else if (thispat->flags == CMD_KEY_READ) {
+                res = sdscatlen(res,"%R~",3);
+            } else if (thispat->flags == CMD_KEY_WRITE) {
+                res = sdscatlen(res,"%W~",3);
```
Comment (hpatro) [inline src/acl.c:771]: As per your initial design, there was a %D  to denote `delete`.

**Review by hpatro — COMMENTED** [review]
Some basic feedback. Can take a deep look after your complete update.

File: src/acl.c (hunk 3 above)
Comment (madolson) [reply src/acl.c:771]: For now I'm thinking about dropping %D. The motivation is that it's hard to document exactly what the difference between write and delete are. R/W also cover most of the interesting cases.

File: src/acl.c (hunk 2 above)
Comment (madolson) [reply src/acl.c:353]: Not sure what you mean by that here?

File: src/acl.c (hunk 1 above)
Comment (madolson) [reply src/acl.c:270]: Users will always have at least one selector, which is always the head node of the list. This allows us to be backwards compatible.

**Pull request closed without merging**